cypress/plugins/synapsedocker/index.ts

```diff
@@ -82,104 +82,126 @@ async function cfgDirFromTemplate(template: string): Promise<SynapseConfig> {
     };
 }
 
+// Start a synapse instance: the template must be the name of
+// one of the templates in the cypress/plugins/synapsedocker/templates
+// directory
+async function synapseStart(template: string): Promise<SynapseInstance> {
+    const synCfg = await cfgDirFromTemplate(template);
+
+    console.log(`Starting synapse with config dir ${synCfg.configDir}...`);
+
+    const synapseId = await new Promise<string>((resolve, reject) => {
+        childProcess.execFile('docker', [
+            "run",
+            "--name", `react-sdk-cypress-synapse-${template}-${randB64Bytes(4)}`,
+            "-d",
+            "-v", `${synCfg.configDir}:/data`,
+            "-p", "8008/tcp",
+            "matrixdotorg/synapse:develop",
+            "run",
+        ], (err, stdout) => {
+            if (err) reject(err);
+            resolve(stdout.trim());
+        });
+    });
+
+    // Get the port that docker allocated: specifying only one
+    // port above leaves docker to just grab a free one, although
+    // in hindsight we need to put the port in public_baseurl in the
+    // config really, so this will probably need changing to use a fixed
+    // / configured port.
+    const port = await new Promise<number>((resolve, reject) => {
+        childProcess.execFile('docker', [
+            "port", synapseId, "8008",
+        ], { encoding: 'utf8' }, (err, stdout) => {
+            if (err) reject(err);
+            resolve(Number(stdout.trim().split(":")[1]));
+        });
+    });
+
+    synapses.set(synapseId, Object.assign({
+        port,
+        synapseId,
+    }, synCfg));
+
+    console.log(`Started synapse with id ${synapseId} on port ${port}.`);
+    return synapses.get(synapseId);
+}
+
+async function synapseStop(id) {
+    const synCfg = synapses.get(id);
+
+    if (!synCfg) throw new Error("Unknown synapse ID");
+
+    const synapseLogsPath = path.join("cypress", "synapselogs", id);
+    await fse.ensureDir(synapseLogsPath);
+
+    await new Promise<void>((resolve, reject) => {
+        childProcess.execFile('docker', [
+            "logs",
+            id,
+        ], async (err, stdout, stderr) => {
+            if (err) reject(err);
+            await fse.writeFile(path.join(synapseLogsPath, "stdout.log"), stdout);
+            await fse.writeFile(path.join(synapseLogsPath, "stderr.log"), stderr);
+            resolve();
+        });
+    });
+
+    await new Promise<void>((resolve, reject) => {
+        childProcess.execFile('docker', [
+            "stop",
+            id,
+        ], err => {
+            if (err) reject(err);
+            resolve();
+        });
+    });
+
+    await new Promise<void>((resolve, reject) => {
+        childProcess.execFile('docker', [
+            "rm",
+            id,
+        ], err => {
+            if (err) reject(err);
+            resolve();
+        });
+    });
```

Comment:
might be best to put this on a `try { } finally { ... }`, but not blocking for this PR

Reply:
👍